### ISSUE TITLE
Failing to read scenario file

### ISSUE THREAD

**Issue by TheNitesWhoSay** [ladislav-zezula/StormLib#293]
StormLib in ReleaseUS compile mode (if not others) fails to read scenario file from attached volcanis.scm map. It is one of the blizzard maps but it seems to open fine in, at least, one of the older Ladik's MPQ editor version (3.5.0.735)/older StormLib (8.20) versions.

[volcanis.zip](https://github.com/ladislav-zezula/StormLib/files/11659501/volcanis.zip)

e.g. following code prints "problem reading file":
```c++
#include <StormLib.h>
#include <cstdint>
#include <iostream>
#include <fstream>
#include <vector>

int main()
{
    HANDLE hMpq = NULL;
    bool opened = SFileOpenArchive(L"C:\\volcanis.scm", NULL, STREAM_FLAG_READ_ONLY, &hMpq);
    size_t bytesRead = 0;
    HANDLE openFile = NULL;
    if ( SFileOpenFileEx(hMpq, "staredit\\scenario.chk", SFILE_OPEN_FROM_MPQ, &openFile) )
    {
        size_t fileSize = (size_t)SFileGetFileSize(openFile, NULL);
        auto fileData = std::vector<std::uint8_t>(fileSize);

        if ( !SFileReadFile(openFile, (LPVOID)&fileData[0], (DWORD)fileSize, (LPDWORD)(&bytesRead), NULL) )
            std::cout << "problem reading file" << std::endl;

        SFileCloseFile(openFile);
    }

    if ( opened )
        SFileCloseArchive(hMpq);

    return 0;
}
```

**Comment by ladislav-zezula** [comment]
Starcraft: Brood War v 1.16 is also unable to read the file. Is there any game version that can actually read it?

**Comment by ladislav-zezula** [comment]
> I believe it's unable to open in StarCraft cause there's a blizzard_2.cnd (blizzard certified) file and the file was resaved, I'll see if I can have this MPQ reproduced without the .cnd file...

This is not true.

In the game:

1) SFileOpenFileEx(hMapMpq, "staredit\\scenario.chk", ..., &hFile) succeeds
2) SFileReadFile(hFile, pbBuffer, 0x24517, &dwBytesRead, NULL) never returns and throws a message box with error about file corruption.

Check this:

* Open the map in MPQ Editor and look to MPQ Tables (Ctrl+T)
* Go to scenario.chk and look at the file sectors.
* The first byte is the compression type
* You can see that first 4 file sectors are compressed with compression 0x01 (Huffmann)
* 5th and subsequent sectors are compressed with 0x03, which would be ZLIB+Huffmann. However, this compression is not allowed in Starcraft, because the entry for ZLIB is not present in the data decompression table:
```
.rdata:15050B4C dcmp_table      dd 40h                  ; DATA XREF: storm_551+15F↑o
.rdata:15050B50                 dd offset Decompress_ADPCM_mono
.rdata:15050B54                 dd 80h
.rdata:15050B58                 dd offset Decompress_ADPCM_stereo
.rdata:15050B5C                 dd 1
.rdata:15050B60                 dd offset Decompress_huff
.rdata:15050B64                 dd 8
.rdata:15050B68                 dd offset Decompress_PKLIB
```
Normally, I would say that someone tried to trick StormLib into a compression that is not supported by the game, but then the game should be able to load the file.

TLDR: This is some sort of f.u.c.k.u.p. file.

**Comment by TheNitesWhoSay** [comment]
Yeah I agree this is an invalid MPQ, I misdiagnosed my issue as a read error (since other tools were still opening it for whatever reason) but it was a write error; this invalid MPQ was created when I (unintentionally) passed HUFFMAN | ZLIB to SFileWriteFile; I consider this my own error and not stormlib's (HUFFMAN shouldn't have been applied to a scenario file in the first place), but maybe generating an invalid MPQ or lacking validation for this flag combo is still not the right behavior; this issue can be closed if you want, otherwise below/attached is my reproduction of the MPQ getting corrupted.

[(2)Volcanis.zip](https://github.com/ladislav-zezula/StormLib/files/11671569/2.Volcanis.zip)

```c++
#include "StormLib.h"
#include <cstdint>
#include <iostream>
#include <fstream>
#include <vector>

int main()
{
    HANDLE hMpq = NULL;
    HANDLE openFile = NULL;
    if ( !SFileOpenArchive(L"C:\\(2)Volcanis.scm", NULL, 0, &hMpq) )
        return 1;
    else if ( !SFileOpenFileEx(hMpq, "staredit\\scenario.chk", SFILE_OPEN_FROM_MPQ, &openFile) )
        return 2;
    
    size_t bytesRead = 0;
    size_t fileSize = (size_t)SFileGetFileSize(openFile, NULL);
    auto fileData = std::vector<std::uint8_t>(fileSize);
    if ( !SFileReadFile(openFile, (LPVOID)&fileData[0], (DWORD)fileSize, (LPDWORD)(&bytesRead), NULL) )
        return 3;
    else if ( bytesRead != fileSize )
        return 4;
    else if ( !SFileCloseFile(openFile) )
        return 5;
    
    openFile = NULL;
    if ( !SFileCreateFile(hMpq, "staredit\\scenario.chk", 0, (DWORD)fileData.size(), 0, MPQ_FILE_COMPRESS | MPQ_FILE_REPLACEEXISTING, &openFile) )
        return 6;
    else if ( !SFileWriteFile(openFile, &fileData[0], (DWORD)fileData.size(), MPQ_COMPRESSION_HUFFMANN | MPQ_COMPRESSION_ZLIB) )
        return GetLastError();
    else if ( !SFileFinishFile(openFile) )
        return 8;
    else if ( !SFileCloseArchive(hMpq) )
        return 9;

    return 0; // Success, but MPQ is corrupted
}
```

**Comment by ladislav-zezula** [comment]
Ok then. My first suspicion was correct then. Closing the issue.

**Comment by ladislav-zezula** [comment]
Fixed. New version of MPQ Editor is up.
